edc-extensions/edr/edr-api-v2/src/main/java/org/eclipse/tractusx/edc/api/edr/v2/EdrCacheApiController.java

```diff
@@ -0,0 +1,194 @@
+                .flatMap(ServiceResult::from)
+                .orElseThrow(exceptionMapper(QuerySpec.class, null)).stream()
+                .map(it -> transformerRegistry.transform(it, JsonObject.class))
+                .peek(r -> r.onFailure(f -> monitor.warning(f.getFailureDetail())))
+                .filter(Result::succeeded)
+                .map(Result::getContent)
+                .collect(toJsonArray());
+    }
+
+    @GET
+    @Path("{transferProcessId}/dataaddress")
+    @Override
+    public JsonObject getEdrEntryDataAddress(@PathParam("transferProcessId") String transferProcessId, @QueryParam("auto_refresh") boolean autoRefresh) {
+
+        var dataAddress = edrStore.resolveByTransferProcess(transferProcessId)
+                .flatMap(ServiceResult::from)
+                .compose(edr -> autoRefresh ? refreshAndUpdateToken(edr, transferProcessId) : success(edr))
+                .orElseThrow(exceptionMapper(EndpointDataReferenceEntry.class, transferProcessId));
+
+        return transformerRegistry.transform(dataAddress, JsonObject.class)
+                .orElseThrow(f -> new EdcException(f.getFailureDetail()));
+
+
+    }
+
+    @DELETE
+    @Path("{transferProcessId}")
+    @Override
+    public void removeEdrEntry(@PathParam("transferProcessId") String transferProcessId) {
+        edrStore.delete(transferProcessId)
+                .flatMap(ServiceResult::from)
+                .orElseThrow(exceptionMapper(EndpointDataReferenceEntry.class, transferProcessId));
+    }
+
+    @POST
+    @Path("{transferProcessId}/refresh")
+    @Override
+    public JsonObject refreshEdr(@PathParam("transferProcessId") String transferProcessId) {
+        var updatedEdr = tokenRefreshHandler.refreshToken(transferProcessId)
+                .orElseThrow(exceptionMapper(DataAddress.class, transferProcessId));
+
+        return transformerRegistry.transform(updatedEdr, JsonObject.class)
+                .orElseThrow(f -> new EdcException(f.getFailureDetail()));
+    }
+
+    // todo: move this method into a service once the "old" EDR api,service,etc. is removed
+    private ServiceResult<DataAddress> refreshAndUpdateToken(DataAddress edr, String id) {
+
+        var edrEntry = edrStore.findById(id);
+        if (edrEntry == null) {
+            return ServiceResult.notFound("An EndpointDataReferenceEntry with ID '%s' does not exist".formatted(id));
+        }
+
+        if (isExpired(edr, edrEntry)) {
+            monitor.debug("Token expired, need to refresh.");
+            return tokenRefreshHandler.refreshToken(id, edr);
+        }
+        return ServiceResult.success(edr);
+    }
+
+    // todo: move this method into a service once the "old" EDR api,service,etc. is removed
+    private boolean isExpired(DataAddress edr, org.eclipse.edc.edr.spi.types.EndpointDataReferenceEntry metadata) {
+        var expiresInString = edr.getStringProperty(TX_AUTH_NS + "expiresIn");
+        if (expiresInString == null) {
```

Comment:
probably if the expiration is not present, we cannot trigger a refresh request? 
it could be a non-refreshable token right?

Reply:
uh... thats right.